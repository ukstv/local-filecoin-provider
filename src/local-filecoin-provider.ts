import type {
  ExtendedKey,
  MessageParams,
  TransactionSignLotusResponse,
} from "@zondax/filecoin-signing-tools";
import type { WalletSubProvider } from "@glif/filecoin-wallet-provider";
import { Network } from "@glif/filecoin-wallet-provider";
import * as bytes from "uint8arrays";

const moduleToImport = process.env.JEST_WORKER_ID
  ? "@zondax/filecoin-signing-tools/nodejs"
  : "@zondax/filecoin-signing-tools";
const signingTools = require(moduleToImport);

export class LocalFilecoinProvider implements WalletSubProvider {
  readonly #privateKey: ExtendedKey;

  constructor(privateKey: string, network: Network = Network.MAIN) {
    const buf = bytes.fromString(privateKey, "base16");
    const json = JSON.parse(bytes.toString(buf));
    const keyType = json.Type;
    const isTestnet = network === Network.TEST;
    if (keyType === "bls") {
      this.#privateKey = signingTools.keyRecoverBLS(json.PrivateKey, isTestnet);
    } else if (keyType === "secp256k1") {
      this.#privateKey = signingTools.keyRecover(json.PrivateKey, isTestnet);
    } else {
      throw new Error(`Unknown key type: ${keyType}`);
    }
  }

  async getAccounts(): Promise<string[]> {
    return [this.#privateKey.address];
  }

  async sign(
    from: string,
    message: MessageParams
  ): Promise<TransactionSignLotusResponse> {
    if (from === this.#privateKey.address) {
      const asString = await signingTools.transactionSignLotus(
        message,
        this.#privateKey.private_base64
      );
      return JSON.parse(asString);
    } else {
      throw new Error(`Can only sign with address ${this.#privateKey.address}`);
    }
  }
}
