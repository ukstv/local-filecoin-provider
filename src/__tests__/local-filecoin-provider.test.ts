import { LocalFilecoinProvider } from "../local-filecoin-provider";

const secp256k1Key =
  "7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a504f3753514931355031343d227d";

describe("secp256k1", () => {
  const provider = new LocalFilecoinProvider(secp256k1Key, true);
  test("#getAccounts", async () => {
    const accounts = await provider.getAccounts();
    expect(accounts).toEqual(["t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka"]);
  });
  test("#sign", async () => {
    const accounts = await provider.getAccounts();
    const address = accounts[0];
    const sig = await provider.sign(address, {
      from: address,
      to: address,
      value: "0",
      method: 0,
      gasPrice: "1",
      gasLimit: 1000,
      nonce: 0,
      params: "",
    });
    expect(sig).toMatchSnapshot();
  });
});
