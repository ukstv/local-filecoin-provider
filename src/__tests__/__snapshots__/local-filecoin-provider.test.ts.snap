// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`secp256k1 #sign 1`] = `
Object {
  "Message": Object {
    "From": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka",
    "GasFeeCap": "1",
    "GasLimit": 1000,
    "GasPremium": "1",
    "Method": 0,
    "Nonce": 0,
    "Params": "",
    "To": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka",
    "Value": "0",
  },
  "Signature": Object {
    "Data": "l3WF6gkTMr1ZD8YUT186/0I7Z+ftoUWs7/2J/vt86qMD5u582Oh+gEcBUq+OaZCNfYdsvq/MTJn8aU07PquT/wE=",
    "Type": 1,
  },
}
`;
